[MERGE] web, web_editor, *: allow image optimize from media dialog

seb-odoo opened the pull request:
Merge commit for PR #31208 

Before this task, many websites were slow due to big images.

Most users upload images without understanding the impact on SEO and speed.

Each upload of an image is important and require some configuration to find the
best solution between speed and quality according to the situation.

After this task, users will have more control on the upload of images,
especially they are given the possibility to reduce their size, quality, and
change the filename.

task-1930726

seb-odoo commented:
Ask discussed with @JKE-be 

- [ ] rename _uploadImageIframe
- [ ] add a note about FormData in the commit https://developer.mozilla.org/en-US/docs/Web/API/FormData
- [ ] double check http route returning json
- [ ] check xhr progress? not sure what this is

seb-odoo commented:
- I renamed the method.
- I double checked the xhr progress, it works nicely with the code I used. (we can consider showing a progress bar to the user if that's something we want functionally)
- I added a note about older browser support in the commit.
- I checked for the http route returning a json, and there is no easy way.

According to this answer: https://stackoverflow.com/a/4083908

> You basically have three choices:
> 
> - Base64 encode the file, at the expense of increasing the data size by around 33%.
> - Send the file first in a multipart/form-data POST, and return an ID to the client. The client then sends the metadata with the ID, and the server re-associates the file and the metadata.
> - Send the metadata first, and return an ID to the client. The client then sends the file with the ID, and the server re-associates the file and the metadata.

So either we keep the current solution: we send the `multipart/form-data` to our 'html' route and return a json.

Or we absolutely want to use RPC + json route for the data, then we need to do either of the above: base64 encode or create 2 routes.

qsm-odoo commented on addons/web_editor/static/src/xml/wysiwyg.xml:
```diff
@@ -99,11 +99,7 @@
-            action="/web_editor/attachment/add"
-            enctype="multipart/form-data"
-            t-att-target="iframeName">
+        <form>
```
I would have left the action attribute here and read it in JS, but it is the same.

JKE-be commented on addons/web_editor/static/src/js/wysiwyg/widgets/alt_dialog.js:
```diff
@@ -2,16 +2,16 @@ odoo.define('wysiwyg.widgets.AltDialog', function (require) {
 
 var core = require('web.core');
-var Dialog = require('wysiwyg.widgets.Dialog');
+var SummernoteDialog = require('wysiwyg.widgets.Dialog');
```
purpose is to remove summernote ;) 
so avoid reference to ;)

JKE-be commented:
@seb-odoo  can we have a PR to review withtout the 9 last commits that are independent and (partially) already merged.

Thank you

seb-odoo commented:
Sure, I'll keep rebasing this. We need to merge (or discard some) the other commits before though because the 3 important commits of the PR depends on them.

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/media.js:
```diff
@@ -335,6 +350,37 @@ var FileWidget = SearchableMediaWidget.extend({
+            }).open();
+        });
+        var always = function () {
+            self.trigger_up('show');
```
rename into show_parent_dialog_request or something like that, those are event name, not names of action to perform

seb-odoo replied:
Ok much better. I wasn't sure if we were supposed to be concise or not. I hate all of those short and barely greppable `show`, `save`, etc...

seb-odoo commented on addons/web_editor/controllers/main.py:
```diff
@@ -1,8 +1,6 @@
 # -*- coding: utf-8 -*-
```
In this case it's a bug introduced in master (image PR).

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/media.js:
```diff
@@ -551,36 +493,72 @@ var FileWidget = SearchableMediaWidget.extend({
+            uploadMutex.exec(function () {
+                return utils.getDataURLFromFile(file).then(function (result) {
+                    return self._rpc({
+                        route: '/web_editor/attachment/add_data',
```
The idea is that if the user selects 10 big files:
- with the mutex: 1st file uploaded alone (10s) -> open optimize modal for this file, other files upload in background, hopefully 2nd upload is done when optimize of 1st file is done, etc.
- without mutex: 10 files uploaded at the same time (100s) -> delays opening the optimize modal at the end, and then nothing happens in background

qsm-odoo replied:
The optimize modal is opened for each file individually in turn ? Why not uploading at the same time (which is less than 100s in your example since this could be done in parallel) and opening the optimize modal for whichever file is ready first (while the others are still being uploaded) ? That way if you have one big files and 9 small files, the modal will open really fast. I think the code would not be more complex (it is just moving the mutex elsewhere I think ?).

seb-odoo replied:
Upload is typically limited by bandwidth, it's hard to get parallel in practice, every additional file uploaded at the same time will delay all the others.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+'use strict';
+
+var core = require('web.core');
+var Dialog = require('web.Dialog');
```
Not needed here, `ImageOptimizeDialog` is only a class to update existing attachments, it has nothing to do with DOM manipulation.

It could be argued that this file should not be on the `wysiwyg` folde then, but where would be best?

qsm-odoo replied:
Mmh, the we dialog was not designed for DOM manipulation but rather to handle a "save" and "cancel" behavior but since you are overridding the button labels anyway... ok. And not it can stay in that folder.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+    /**
+     * @constructor
+     */
+    init: function (parent, options, params) {
```
The goal was to split the options that go to the parent dialog and those that are used by the optimize dialog itself... Not sure the best way to handle it or if that's something we do at all, but I don't like when they are mixed up.

qsm-odoo replied:
Mmh. params first at least since when you instantiate your dialog, you are more likely to want to pass params than options then.

seb-odoo replied:
done

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+     * the image width (we don't scale up images).
+     *
+     * @private
+     * @param {number} size: integer
```
So in this case we don't care that `integer` is not an actual type?

qsm-odoo replied:
You don't write "object" for promises and jquery element right ? ;)

seb-odoo replied:
For objects we write the name of the "class" when there is one, that makes sense to me. But I'm ok either way for int, just asking.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/media.js:
```diff
@@ -362,43 +349,21 @@ var FileWidget = SearchableMediaWidget.extend({
-            .groupBy(function (a, index) { return Math.floor(index / self.IMAGES_PER_ROW); })
-            .values()
-            .value();
+        var attachments = _(this.attachments).slice(0, this.numberOfAttachmentsToDisplay);
```
This is how it was done before, but sure I'll change it to array slice

qsm-odoo replied:
Yes, but since you are touching the line anyway... :)

seb-odoo commented on addons/web_editor/static/src/scss/wysiwyg.scss:
```diff
@@ -287,49 +287,62 @@ body .modal {
+                    }
+                    &.o_we_attachment_selected {
+                        border-color: $o-brand-primary;
+                        box-shadow: 0px 0px 2px 2px $o-brand-primary;
```
The problem here is that the DOM is different, for the image case we want to put the border around the image `img`, but for the document we want the border around the whole area and not just around the small image.

The current diff is actually bad because I forgot for some rules to remove the `img` from my copy/paste into documents.

How to handle it then?

qsm-odoo replied:
From worst to best:
- Using a scss mixin
- Using an extend of scss placeholder
- Reviewing the DOM to add the targetable class on the img/the container depending on the case

seb-odoo replied:
Done with best option.

seb-odoo commented on addons/web_editor/static/tests/wysiwyg_snippets_tests.js:
```diff
@@ -330,6 +330,12 @@ QUnit.test('background-image', function (assert) {
                 '</section>',
         },
+        mockRPC: function (route, args) {
+            if (route === '/web_editor/static/src/img/transparent.png?height=192') {
```
Yes I will gladly do that. Wasted so much time on this.

seb-odoo commented on addons/web_editor/models/ir_attachment.py:
```diff
@@ -9,10 +11,29 @@ class IrAttachment(models.Model):
             self.local_url = '/web/image/%s?unique=%s' % (self.id, self.checksum)
+
+    @api.depends('mimetype', 'url', 'name', 'datas_fname')
+    @api.multi
```
Dunno if it makes any difference but OK to change.

qsm-odoo replied:
Difference in my sanity level :3

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+var ImageOptimizeDialog = Dialog.extend({
+    template: 'wysiwyg.widgets.image.preview',
+    xmlDependencies: ['/web_editor/static/src/xml/wysiwyg.xml'],
+
```
ok

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+        'input #o_we_width': '_onWidthInput',
+        'input .js_quality_range': '_onQualityRangeInput',
+    }),
+    /**
```
Looks like a messed up rebase, changed.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+
+var ImageOptimizeDialog = Dialog.extend({
+    template: 'wysiwyg.widgets.image.preview',
+    xmlDependencies: ['/web_editor/static/src/xml/wysiwyg.xml'],
```
I should know... changed.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+        this.isExisting = params.isExisting;
+        this.attachment = params.attachment;
+        this.resolve = params.resolve;
+        this.reject = params.reject;
```
Just leaving a comment to not forget, we need to discuss about this today so I can do it properly when you are not there.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+            this.attachment.id, width, height, quality));
+        this.$widthPresets.removeClass('active');
+        this.$widthPresets.each(function () {
+            var $button = $(this);
```
ok

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+     *
+     * @private
+     */
+    _onOptimize: function (ev) {
```
ok

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+        var self = this;
+        var filename = this.$filename.val();
+        if (!filename || this.$widthInput.val() > this.attachment.image_width || this.$widthInput.val() < 0) {
+            return Promise.reject();
```
Agree, I will disable the button when there is wrong input.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+        var params = {
+            'filename': filename,
+        };
+        params['quality'] = parseInt(this.$qualityRange.val());
```
It was on a `if` at some point during dev, but indeed now it can be moved up.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+        if (!quality) {
+            return;
+        }
+        this.$qualityRange.val(quality).trigger('input');
```
I didn't know what is the guideline for this, but yes here the function could be called directly since the `ev` is not even accessed at all.

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,223 @@
+        ev.preventDefault();
+        this.$widthInput.val(parseInt($(ev.target).data('width'))).trigger('input');
+    },
+
```
ok

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/media.js:
```diff
@@ -1,19 +1,20 @@
-
 var _t = core._t;
 
+var ImageOptimizeDialog = require('wysiwyg.widgets.image_optimize_dialog').ImageOptimizeDialog;
```
ok

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/media.js:
```diff
@@ -258,6 +270,9 @@ var FileWidget = SearchableMediaWidget.extend({
             .replace(allImgClasses, ' ')
             .replace(allImgClassModifiers, ' ');
     },
+    _computeOptimizedWidth: function () {
```
added

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/media.js:
```diff
@@ -362,43 +349,21 @@ var FileWidget = SearchableMediaWidget.extend({
-        if (withEffect) {
-            Promise.all(imageDefs).then(function () {
-                _.delay(function () {
-                    $divs.removeClass('o_image_loading');
```
As discussed, I will remove it (properly).

seb-odoo commented:
Rebase done (no conflict)

seb-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_optimize_dialog.js:
```diff
@@ -0,0 +1,304 @@
+        //  - SVG because it doesn't make sense
+        //  - GIF because our current code is not made to handle all the frames
+        this.disableResize = ['image/jpeg', 'image/jpe', 'image/jpg', 'image/png'].indexOf(this.attachment.mimetype) === -1;
+        this.disableQuality = ['image/jpeg', 'image/jpe', 'image/jpg', 'image/png'].indexOf(this.attachment.mimetype) === -1;
```
done

JKE-be commented on addons/web_editor/models/ir_attachment.py:
```diff
@@ -1,18 +1,53 @@
+    @api.depends('mimetype', 'url', 'name', 'datas_fname')
+    def _compute_image_src(self):
+        for attachment in self:
+            if not any(m == attachment.mimetype for m in ['image/gif', 'image/jpe', 'image/jpeg', 'image/jpg', 'image/gif', 'image/png', 'image/svg+xml']):
```
attachment.mimetype in ['image/gif', 'image/jpe', 'image/jpeg', 'image/jpg', 'image/gif', 'image/png', 'image/svg+xml']

seb-odoo replied:
Oops... changed

seb-odoo commented:
and removed waiting forward port commits as they don't directly impact the new feature

seb-odoo commented:
@JKE-be all changes done, see last 2 force push:

- removed suggested size 64
- removed datas_fname (note: this will break filters background_ if the user changes the name to remove it)
- changed optimize label for png + added help text
- changed help to not contain website
- removed inline style and adapted some class names
- changed icon for quick upload to fast-forward and hide it in debug

JKE-be commented:
@robodoo r+

JKE-be commented:
@robodoo merge

robodoo commented:
Merge method set to merge directly, using the PR as merge commit message

seb-odoo commented:
Ca serait mieux rebase merge non ?

Le mar. 4 juin 2019 à 19:55, Robot Odoo <notifications@github.com> a écrit :

> Merge method set to merge directly, using the PR as merge commit message
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/odoo/odoo/pull/31208?email_source=notifications&email_token=AJDRPAGF36T4RROZO77DB3DPY2UCTA5CNFSM4GYD76EKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODW5LTTY#issuecomment-498776527>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJDRPADWK7BFXBXHDSCUNHTPY2UCTANCNFSM4GYD76EA>
> .
>

robodoo commented:
Linked pull request(s) odoo/enterprise#4459 not ready. Linked PRs are not staged until all of them are ready.

qsm-odoo commented:
@robodoo rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

JKE-be commented:
@robodoo merge

robodoo commented:
Merge method set to merge directly, using the PR as merge commit message

JKE-be commented:
@robodoo r+ p=1

robodoo commented:
Merged, thanks!
